### PR TITLE
CIP-30: BLS12-377 Precompiles

### PR DESCRIPTION
**Pull request by kilic**
### Description

This pr adds BLS12-377 elliptic curve operations as precompiles. Precompiles are added in same fashion with [EIP-2537](https://eips.ethereum.org/EIPS/eip-2537) except the mapping functions.

#### Background

Elliptic curve library is adapted from [BLS12-381](https://github.com/kilic/bls12-381) and [EIP-1962](https://github.com/saitima/eip1962) works. x86 optimised field implementation is generated with [fp](https://github.com/kilic/fp) and native go implementation is generated with [goff](https://github.com/ConsenSys/goff)

### Other changes

Gas table for these new precompiles set is not decided yet. Rather than actual numbers placeholder constants are placed.

### Tests

Underlying elliptic curve library is heavily tested in terms of mathematical properties. 

Original library BLS12-381 is subjected to the [security audit](https://github.com/kilic/bls12-381/issues/19) under [drand project](https://drand.love/blog/2020/08/10/drand-launches-v1-0/) and also made its way to be backend of EIP-2537 go-ethereum precompiles[*](https://github.com/ethereum/go-ethereum/pull/21018) for the next HF.

Cross and fuzz testing are not applied for the adapted library yet, also _static call_ test vectors are not generated.

### Backwards compatibility

Precompiles are included next to Celo specific precompile bundle. It is assumed that precompile addresses incrementally follow the others.

### CODE REVIEW

**Comment by prestwich** [comment]
There are pricings specified in [EIP-2539](https://github.com/ethereum/EIPs/blob/master/EIPS/eip-2539.md) based on benching the 1962 library. They are precisely equal to the [EIP-2537](https://github.com/ethereum/EIPs/blob/master/EIPS/eip-2537.md) prices, so we should be able to re-use that pricing code for this PR as well.

I would prefer if this were explicitly an implementation of EIP-2539 (besides any address changes)

**Comment by prestwich** [comment]
Update: I talked to @shamatar and he identified a metering issue that will result in gas price changes and a slight divergence between the EIPs. So hold off on making gas changes until those come back

**Comment by kilic** [comment]
Zexe test vectors are generated with https://github.com/kilic/zexe/blob/eip2539/algebra/src/bls12_377/curves/eip2539_test.rs

@prestwich we may want to reduce scalar multiplication gas price in short term since GLV multiplication which is more efficient is planned to be implemented.

**Comment by kilic** [comment]
Matter test vectors are generated with https://github.com/kilic/eip1962-1/tree/celo1157/src/public_interface/celo1157

However there are some issues of matter eip1962 implementation.

1. b coefficient of G1 and G2 were wrong (I have fixed for test gens)
2. Pairing equation outputs 1 where it should output 0. [Here](https://github.com/kilic/eip1962-1/blob/65bc07b37af300dcd4bf9a87961b0d7c0f4cb1d6/src/public_interface/celo1157/mod.rs#L1208), apparently for odd numbers we expect 0. So I have changed manually to 0 which is I think not appropriate bad to do so. I haven't got a chance to what is wrong with it yet.

**Comment by prestwich** [comment]
See gas metering update PR here:
https://github.com/ethereum/EIPs/pull/2999

**Comment by prestwich** [comment]
> * b coefficient of G1 and G2 were wrong (I have fixed for test gens)
> * Pairing equation outputs 1 where it should output 0. [Here](https://github.com/kilic/eip1962-1/blob/65bc07b37af300dcd4bf9a87961b0d7c0f4cb1d6/src/public_interface/celo1157/mod.rs#L1208), apparently for odd numbers we expect 0. So I have changed manually to 0 which is I think not appropriate bad to do so. I haven't got a chance to what is wrong with it yet.

My open PR in 1962 addresses the b coefficient issue https://github.com/matter-labs/eip1962/pull/14

I'll notify the Matter team and see if they'll take a look at the pairing output issue

**Comment by shamatar** [comment]
> 2. Pairing equation outputs 1 where it should output 0. [Here](https://github.com/kilic/eip1962-1/blob/65bc07b37af300dcd4bf9a87961b0d7c0f4cb1d6/src/public_interface/celo1157/mod.rs#L1208), apparently for odd numbers we expect 0. So I have changed manually to 0 which is I think not appropriate bad to do so. I haven't got a chance to what is wrong with it yet.

I've also updated coefficients in my optimizations branch (that uses special forms of non-residues), but what is a problem with a line there exactly? If number of pairs is even we should get combinations like `e(a, b) * e(-a, b) = 1`

File: core/vm/contracts.go
Comment (prestwich) [inline core/vm/contracts.go:79]: I am still optimistic that we can include 2539 in Berlin, in which case these would change

File: core/vm/contracts.go
Comment (prestwich) [inline core/vm/contracts.go:1090]: should be `EIP-2539` throughout :)

**Comment by prestwich** [comment]
I've been running some compatibility tests between this work and the rust eip1962 lib. I'm getting a mismatch on the pairing function with a specific input (below). The Golang implementation (accessed via geth rpc) returns `false` while the rustlang returns `true`. Given that these points were selected randomly, I would expect the output to be `false`.

I am working on this branch of eip1962:
https://github.com/prestwich/eip1962/tree/prestwich/eip2539

```
000000000000000000000000000000000096ad951b95fb9160ed9087c35f5eab95e1e8aefe47440d7f1f1ec5cfde6f7a6caa774caca46ba2d988b16fac2831ff0000000000000000000000000000000000467e7134a6743759be14b955e7647983308dc12a389ca2fa3084ab54750afc82d7c4a58e5158aa307e0a6810c293a100000000000000000000000000000000015afe67f1ebc55d6bc08101957e673d56a5e675191f191ee537dbd3adaf342d3392147ab369b2cd930157dae07e178c0000000000000000000000000000000001511a4c76001b4acb211d77c5768da96c4b991075508065d8485ddda6fd4cf254f27e417f5d5da7234e6f48de449c39000000000000000000000000000000000075e634a2cfd17d1eb9b42ee6bee979e6cfbbb1a73608431fb8f783382b5a2c876580821a944c32b6637022a3d910f30000000000000000000000000000000000ab07f36540bb3ccad93061a48c5cc8848b5418e6b2eca075c439c2b336df56635552f9019980c58713fcfd74acab39
```

**Comment by kilic** [comment]
@prestwich 

> I would appreciate someone else checking this out and telling me if it's likely to be an error in my testing setup while I try to debug locally

@shamatar 

> I've also updated coefficients in my optimizations branch (that uses special forms of non-residues), but what is a problem with a line there exactly?

I have added a small test below that demonstrates the issue.

https://github.com/kilic/eip1962-1/blob/1a8292cf676797025960b8d9b3328168394d3be5/src/public_interface/celo1157/mod.rs#L1060

**Comment by shamatar** [comment]
> @prestwich
> 
> Did mismatch happen in the very first call or in a kind of fuzzing attempt?
> 
> > I would appreciate someone else checking this out and telling me if it's likely to be an error in my testing setup while I try to debug locally
> 
> @shamatar
> 
> > I've also updated coefficients in my optimizations branch (that uses special forms of non-residues), but what is a problem with a line there exactly?
> 
> I have added a small test below that demonstrates the issue.
> 
> https://github.com/kilic/eip1962-1/blob/1a8292cf676797025960b8d9b3328168394d3be5/src/public_interface/celo1157/mod.rs#L1060

@kilic Should I run this test for BLS12-377 curve? (to double check)

**Comment by shamatar** [comment]
Fixed the hardcoded parameters. Please use the "tuning" branch for now, it contains an API implementation for 2539, so can be tested from outside (signatures should match)

**Comment by prestwich** [comment]
> Did mismatch happen in the very first call or in a kind of fuzzing attempt?

On the very first call. I've also been running the others with random inputs generated by the eip1962 library without issues

**Comment by prestwich** [comment]
> Fixed the hardcoded parameters. Please use the "tuning" branch for now, it contains an API implementation for 2539, so can be tested from outside (signatures should match)

I will update my work to use the `tuning` branch and get back to you.

**Comment by shamatar** [comment]
There is a demo how to build generic fuzzer for precompiles like 2537/2539 in this repo https://github.com/shamatar/algebraic_fuzzer

It does not perform any actual comparison, just fuzzes G1 addition for now, but it looks to be ok to template it over some form of generators and runners

**Comment by shamatar** [comment]
There is now also branch with square root functions for 1 mod 16 (case of base field of BLS12-377), so we can actually fix a fuzzer for points in invalid subgroup and not on curve

**Comment by prestwich** [comment]
My eip2539 branch of the fuzzer has been updated here: https://github.com/prestwich/algebraic_fuzzer

We are provisioning a node to run it for a bit.

File: core/vm/contracts.go
Comment (prestwich) [reply core/vm/contracts.go:79]: I am no longer optimistic.

**Comment by prestwich** [comment]
> My eip2539 branch of the fuzzer has been updated here: https://github.com/prestwich/algebraic_fuzzer
> 
> We are provisioning a node to run it for a bit.

Ran about 1billion iterations of EIP-2539 and 1.5 billion of EIP-2537. No issues encountered.

Given we have tests and a large number of fuzzing runs against an integration, we should update the gas costs to match the latest EIP2539. Any other things that need to be done before this is ready for review for inclusion in Donut?

File: params/protocol_params.go
Comment (prestwich) [inline params/protocol_params.go:156]: we can fill these from [CIP-30](https://github.com/celo-org/celo-proposals/blob/master/CIPs/cip-0030.md). I think after Donut we can re-examine pricing to align to recently improved benchmarks :)

**Review by prestwich — CHANGES_REQUESTED** [review]

**Comment by prestwich** [comment]
hey @kilic, I'm starting to shephard client integration. :) Can you please do the following:

1. once #1286 is merged, rebase onto latest master
2. Remove your current address declarations
3. Remove any activation logic (including any rows in the istanbul precompile block)
4. Fill in the rows row in the Donut Precompile block

File: params/protocol_params.go
Comment (prestwich) [inline params/protocol_params.go:152]: We still need to fill these in

**Comment by mcortesi** [comment]
Closed in favor of #1341

**Pull request closed without merging**
